enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -2435,12 +2453,129 @@ func newPartitionEvictor(part disk.Partition, fileStorer filestore.Store, blobDi
+	})
+	eg.Wait()
+	// Drain samples chan before exiting
+	for len(e.samples) > 0 {
+		<-e.samples
+	}
+	close(e.samples)
+}
+
+func (e *partitionEvictor) generateSamplesForEviction(quitChan chan struct{}) error {
+	db, err := e.dbGetter.DB()
+	if err != nil {
+		log.Warningf("cannot generate samples for eviction: failed to get db: %s", err)
+		return err
+	}
+	defer db.Close()
+	start, end := keyRange([]byte(e.partitionKeyPrefix() + "/"))
+	iter := db.NewIter(&pebble.IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+	})
+	// We update the iter variable later on, so we need to wrap the Close call
+	// in a func to operate on the correct iterator instance.
+	defer func() {
+		iter.Close()
+	}()
+
+	totalCount := 0
+	shouldCreateNewIter := true
+
+	// Files are kept in random order (because they are keyed by digest), so
+	// instead of doing a new seek for every random sample we will seek once
+	// and just read forward, yielding digests until we've found enough.
+
+	for {
+		select {
+		case <-quitChan:
+			return nil
+		default:
+		}
+		// Refresh the iterator once a while
+		if shouldCreateNewIter {
+			shouldCreateNewIter = false
+			totalCount = 0
+			newIter := db.NewIter(&pebble.IterOptions{
+				LowerBound: start,
+				UpperBound: end,
+			})
+			iter.Close()
+			iter = newIter
+		}
+		totalCount += 1
+		if totalCount > e.samplesPerBatch {
+			// Going to refresh the iterator in the next iteration.
+			shouldCreateNewIter = true
+		}
+		if !iter.Valid() {
+			// This should happen once every totalCount times or when
+			// we exausted the iter.
+			randomKey, err := e.randomKey(64)
+			if err != nil {
+				log.Warningf("cannot generate samples for eviction: failed to get random key: %s", err)
+				return err
+			}
+			valid := iter.SeekGE(randomKey)
+			if !valid {
+				shouldCreateNewIter = true
+				continue
+			}
+		}
+		var key filestore.PebbleKey
+		if _, err := key.FromBytes(iter.Key()); err != nil {
+			log.Warningf("cannot generate sample for eviction, skipping: failed to read key: %s", err)
+			continue
+		}
+
+		fileMetadata := &rfpb.FileMetadata{}
+		err = proto.Unmarshal(iter.Value(), fileMetadata)
+		if err != nil {
+			log.Warningf("cannot generate sample for eviction, skipping: failed to read proto: %s", err)
+			continue
+		}
+
+		atime := time.UnixMicro(fileMetadata.GetLastAccessUsec())
+		age := e.clock.Since(atime)
+
+		sizeBytes := fileMetadata.GetStoredSizeBytes()
+		if e.includeMetadataSize {
+			sizeBytes = getTotalSizeBytes(fileMetadata) + int64(len(iter.Key()))
+		}
+
+		if age >= e.minEvictionAge {
+			keyBytes := make([]byte, len(iter.Key()))
+			copy(keyBytes, iter.Key())
+			sample := &approxlru.Sample[*evictionKey]{
+				Key: &evictionKey{
+					bytes:           keyBytes,
+					storageMetadata: fileMetadata.GetStorageMetadata(),
+				},
+				SizeBytes: sizeBytes,
+				Timestamp: atime,
+			}
+			select {
+			case e.samples <- sample:
+			case <-quitChan:
+				return nil
+			}
+		}
+		iter.Next()
```

Comment:
could you have the outer for be:

`for iter.Next() {`  and remove this one?

Reply:
No I can't. Even when iter.Next() is invalid, I still want to be in the loop: create new random key or create new iterators.